esp-hal/src/dma/mod.rs

```diff
@@ -902,141 +953,296 @@ impl DescriptorChain {
             return Err(DmaError::UnsupportedMemoryRegion);
         }
 
-        if self.descriptors.len() < len.div_ceil(self.chunk_size) {
-            return Err(DmaError::OutOfDescriptors);
-        }
+        let max_chunk_size = self.max_chunk_size(len, circular)?;
 
-        if circular && len <= 3 {
-            return Err(DmaError::BufferTooSmall);
-        }
+        DescriptorSet::link_with_buffer_impl(
+            unsafe { core::slice::from_raw_parts_mut(data, len) },
+            self.descriptors,
+            max_chunk_size,
+            circular,
+        )?;
+        DescriptorSet::prepare_descriptors_impl(
+            self.descriptors,
+            len,
+            max_chunk_size,
+            circular,
+            prepare_descriptor,
+        )?;
 
-        self.descriptors.fill(DmaDescriptor::EMPTY);
+        Ok(())
+    }
 
-        let max_chunk_size = if !circular || len > self.chunk_size * 2 {
-            self.chunk_size
-        } else {
+    fn max_chunk_size(&self, len: usize, circular: bool) -> Result<usize, DmaError> {
+        let max_chunk_size = if circular && len <= self.chunk_size * 2 {
+            if len <= 3 {
+                return Err(DmaError::BufferTooSmall);
+            }
             len / 3 + len % 3
+        } else {
+            self.chunk_size
         };
 
-        let mut processed = 0;
-        let mut descr = 0;
-        loop {
-            let chunk_size = usize::min(max_chunk_size, len - processed);
-            let last = processed + chunk_size >= len;
+        Ok(max_chunk_size)
+    }
+}
 
-            let next = if last {
-                if circular {
-                    addr_of_mut!(self.descriptors[0])
-                } else {
-                    core::ptr::null_mut()
-                }
-            } else {
-                addr_of_mut!(self.descriptors[descr + 1])
-            };
+#[doc(hidden)]
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct DescriptorSet<'a> {
+    descriptors: &'a mut [DmaDescriptor],
+    block_size: Option<DmaBufBlkSize>,
```

Comment:
Does this need to be stored in here? It's not always used, i.e. `DmaRxBuf` and `DmaRxTxBuf` don't use it.

Reply:
I imagine that will change once we can read into PSRAM, right?

Reply:
I suppose but the field would be duplicated in the `DmaRxTxBuf` case and it's a bit weird to seeing the field in here without the additional alignment checks required for RX.